### PR TITLE
Support Microsoft Entra Id authentication for rest source

### PR DESCRIPTION
**Pull request by yao-msft**
## Change

This pr adds Microsoft Entra Id support for rest source. To achieve this, rest source /information contract is updated with below. The /information endpoint should never use authentication as it is the first handshake between winget client and rest source.

```
// "Authentication": {
//     "AuthenticationType": "microsoftEntraId",
//     "MicrosoftEntraIdAuthenticationInfo" : {
//         "Resource": "Resource",
//         "Scope" : "test"
//     }
// }
```
Rest source interface will read this info and acquire authentication using WebAccountManager OS api during all following source interaction (/packageManifests, /manifestSearch)

CLI arguments updated to specify authentication behavior

COM interface updated to specify authentication behavior from caller and provide authentication info of package catalog to caller

## Validation

- Due to lack of e2e rest source tests and complex in getting real Microsoft Entra Id token. E2e validation is manually performed. I tested cli (packaged and unpackaged), InProcCom and OutOfProcCom . Both interactive and silent flow works correctly. I setup a test rest source with Microsoft Entra Id protection (based on tenants) and it's working as expected
- Added a bunch of unit tests: AuthenticationInfo parsing, RestClient and RestInterface creation, and RestInterface GetManifest and Search operation with override Microsoft Entra Id token

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/winget-cli/pull/4123)

### CODE REVIEW

**Comment by denelon** [comment]
* https://github.com/microsoft/winget-cli/issues/4124

File: src/AppInstallerCommonCore/Authentication/WebAccountManagerAuthenticator.cpp [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+            return result;
+        }
+
+        if (requestResult.ResponseStatus() == WebTokenRequestStatus::Success)
+        {
+            auto responseData = requestResult.ResponseData().GetAt(0);
+            auto authenticatedAccount = responseData.WebAccount();
+
+            // Check token's corresponding account matches user input if applicable.
+            if (m_authArgs.AuthenticationAccount.empty() || Utility::CaseInsensitiveEquals(m_authArgs.AuthenticationAccount, Utility::ConvertToUTF8(authenticatedAccount.UserName())))
```
Comment (ryfu-msft) [inline src/AppInstallerCommonCore/Authentication/WebAccountManagerAuthenticator.cpp:242]: >m_authArgs.AuthenticationAccount.empty() [](http://example.com/codeflow?start=16&length=40)

Does an empty authenticationAccount mean that it succeeded?

**Review by ryfu-msft — APPROVED** [review]

File: src/AppInstallerCommonCore/Authentication/WebAccountManagerAuthenticator.cpp (hunk 1 above)
Comment (yao-msft) [reply src/AppInstallerCommonCore/Authentication/WebAccountManagerAuthenticator.cpp:242]: Empty m_authArgs.AuthenticationAccount means user or caller did not specify which account to use. So we consider it as authentication success as long as WebAccountmanager returns success.

**Pull request merged**